pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/RealtimeSegmentAssignment.java

```diff
@@ -365,12 +405,32 @@ private List<String> assignCompletedSegment(String segmentName, Map<String, Map<
       // Replica-group based assignment
 
       // Uniformly spray the segment partitions over the instance partitions
-      int segmentPartitionId =
-          SegmentUtils.getRealtimeSegmentPartitionId(segmentName, _realtimeTableName, _helixManager, _partitionColumn);
-      int numPartitions = instancePartitions.getNumPartitions();
-      int partitionGroupId = segmentPartitionId % numPartitions;
+      int desiredPartitionId = getDesiredPartitionId(segmentName, instancePartitions, currentAssignment);
       return SegmentAssignmentUtils
-          .assignSegmentWithReplicaGroup(currentAssignment, instancePartitions, partitionGroupId);
+          .assignSegmentWithReplicaGroup(currentAssignment, instancePartitions, desiredPartitionId);
     }
   }
+
+  private int getDesiredPartitionId(String segmentName, InstancePartitions instancePartitions,
+      Map<String, Map<String, String>> currentAssignment) {
+    Optional<Integer> segmentPartitionIdOpt =
+        SegmentUtils.getRealtimeSegmentPartitionId(segmentName, _realtimeTableName, _helixManager, _partitionColumn);
+    int numPartitions = instancePartitions.getNumPartitions();
+    return segmentPartitionIdOpt.map(segPartitionId -> segPartitionId % numPartitions).orElseGet(() -> {
```

Comment:
Suggest not using functional apis. Even though this method is not very performance critical, it is still called very frequently when rebalancing the table